File: src/hyp_object.rs
//! Hyperbolic Geometry
//!
//! This module provides support for hyperbolic geometry using the Cayley-Klein
//! model. In hyperbolic geometry, the perpendicular relationship is defined by the
//! form x² + y² - z² = 0.

use crate::pg_object::{HyperbolicLine, HyperbolicPoint};
use crate::{CayleyKleinPlane, CayleyKleinPlanePrimitive};

/// Coefficients for perpendicular transformation in hyperbolic geometry.
/// These define the polarity for the hyperbolic metric.
const HYP_PERP_COEFFS: [i64; 3] = [1, 1, -1];

impl_cayley_klein_plane!(
    HyperbolicPoint,
    HyperbolicLine,
    |p: &HyperbolicPoint| {
        HyperbolicLine::new([
            HYP_PERP_COEFFS[0] * p.coord[0],
            HYP_PERP_COEFFS[1] * p.coord[1],
            HYP_PERP_COEFFS[2] * p.coord[2],
        ])
    },
    |l: &HyperbolicLine| {
        HyperbolicPoint::new([
            HYP_PERP_COEFFS[0] * l.coord[0],
            HYP_PERP_COEFFS[1] * l.coord[1],
            HYP_PERP_COEFFS[2] * l.coord[2],
        ])
    }
);

File: python_ai/src/projgeom/hyp_object.py
"""Hyperbolic Geometry Objects module.

This module defines hyperbolic geometry points and lines with their
perpendicularity operations.
"""

from .ck_plane import CayleyKleinPlane
from .pg_object import HyperbolicLine, HyperbolicPoint

# Hyperbolic perpendicularity coefficients
HYP_PERP_COEFFS = [1, 1, -1]


def _perp_point(point: HyperbolicPoint) -> HyperbolicLine:
    """Return the polar line of a hyperbolic point."""
    return HyperbolicLine(
        [
            HYP_PERP_COEFFS[0] * point.coord[0],
            HYP_PERP_COEFFS[1] * point.coord[1],
            HYP_PERP_COEFFS[2] * point.coord[2],
        ]
    )


def _perp_line(line: HyperbolicLine) -> HyperbolicPoint:
    """Return the pole of a hyperbolic line."""
    return HyperbolicPoint(
        [
            HYP_PERP_COEFFS[0] * line.coord[0],
            HYP_PERP_COEFFS[1] * line.coord[1],
            HYP_PERP_COEFFS[2] * line.coord[2],
        ]
    )


# Implement CayleyKleinPlanePrimitive for HyperbolicPoint
HyperbolicPoint.perp = _perp_point  # type: ignore[attr-defined]

# Implement CayleyKleinPlanePrimitive for HyperbolicLine
HyperbolicLine.perp = _perp_line  # type: ignore[attr-defined]

# Implement CayleyKleinPlane for HyperbolicPoint and HyperbolicLine
CayleyKleinPlane.register(HyperbolicPoint)
CayleyKleinPlane.register(HyperbolicLine)
